Abbr Extension: Definition Sorting and Glossary storage

nbanyan opened the pull request:
Several changes/features here.  All input welcome on each.  

* **Sorting**: A simple sort by length to process longer abbreviations first solves all instances of smaller abbreviations preventing longer ones from being used.
* `glossary`: An option to include a Markdown file with abbreviations to apply to every page.
* `use_last_abbr`: An option to only keep the first instance of an abbreviation (useful when using other methods to append glossaries to pages).

The glossary ended up being the more complex of the three and involved replicating some of the code/process from **core.py**.

nbanyan committed:
`AbbrTreeprocessor` now sorts the abbreviation
list by length before processing the element tree

This ensures that multi-word abbreviations are
implemented even if an abbreviation exists for
one of those component words.

To test this,
added `test_abbr_superset_vs_subset()` to
`tests.test_syntax.extensions.test_abbr`.

nbanyan committed:
Added a configuration option to `abbr` and
handling for empty abbreviations.

- Configuration option `use_last_abbr` allows using
the first instance of a term rather than the last.
This is useful for sites that auto-append a glossary
to pages.
- If an abbreviation does not have a definition it
will skip the term instead of writing abbr tags with
an empty title.

nbanyan committed:
Added a `glossary` option for the `abbr` extension

The glossary file also uses the Markdown
abbreviation syntax (`AbbrBlockprocessor` is
used to process the file) and keeps the glossary
definitions separate from the page definitions,
allowing the glossary to be applied to every page
while only being processed once.

waylan commented:
So far I've only taken a cursory look at your changes. Here are my initial thoughts.

> * **Sorting**: A simple sort by length to process longer abbreviations first solves all instances of smaller abbreviations preventing longer ones from being used.

This is good. I should have caught the issue this fixes with the recent refactor. Thank you.

> * `glossary`: An option to include a Markdown file with abbreviations to apply to every page.

I'm not sure I am comfortable with this. In our discussion in #1465 I suggested allowing a user to pass in a Python object (such as a dict), not importing a file of Markdown content. I'm inclined to reject this on the basis that extensions already exist which allow Markdown content to be imported.

> * `use_last_abbr`: An option to only keep the first instance of an abbreviation (useful when using other methods to append glossaries to pages).

As previously explained, I don't see any value in this. You still need to ensure your definitions are in a certain order. That order is now reversed which is counterintuitive (IMO) and works the opposite of every other reference-type syntax (reference links, footnotes, etc) in Markdown. I realize this is an option which if off by default, but with there being no value to changing the setting, then why even offer it? That said, if you can provide a concrete example which demonstrates its value and changes my mind, then I would suggest renaming it `use_first_abbr` with the default being `False`. That name more clearly aligns with the description of the option.

nbanyan committed:
Changed Abbreviation Extension's `glossary` option
to accept a dictionary instead of a file.

Changed `load_glossary` to merge in a dictionary,
overwriting any existing records.

Added a `reset_glossary` method to clear the glossary.

Removed the `use_last_abbr` option, since `glossary`
can be used to resolve the use case `use_last_abbr`
was written for.

nbanyan commented:
Discarded `use_last_abbr` and changed `glossary` to accept and manage dictionary objects directly.

waylan commented on docs/changelog.md:
```diff
@@ -23,6 +23,24 @@ better reflects what it is. `AbbrPreprocessor` has been deprecated.
+to ensure that multi-word abbreviations are implemented even if an abbreviation
+exists for one of those component words. (#1465)
+
+Empty abbreviations are now skipped by `AbbrTreeprocessor`. This avoids applying
```
Changelog needs to be updated to reflect latest changes.

nbanyan replied:
Updated.

waylan commented on docs/extensions/abbreviations.md:
```diff
@@ -46,7 +51,20 @@ Usage
-This extension does not accept any special configuration options.
+The following options are provided to configure the output:
+
+* **`use_last_abbr`**:
```
Docs need updated to reflect latest changes.

nbanyan replied:
Updated. Let me know if more details are needed for the `glossary` option and functions.

waylan commented on markdown/extensions/abbr.py:
```diff
@@ -23,6 +23,7 @@
 from __future__ import annotations
 
 from . import Extension
+from ..util import parseBoolValue
```
It looks like this is not called and can be removed.

waylan commented on markdown/extensions/abbr.py:
```diff
@@ -41,18 +42,41 @@ class AbbrExtension(Extension):
         md.registerExtension(self)
         md.treeprocessors.register(AbbrTreeprocessor(md, self.abbrs), 'abbr', 7)
-        md.parser.blockprocessors.register(AbbrBlockprocessor(md.parser, self.abbrs), 'abbr', 16)
+        md.parser.blockprocessors.register(AbbrBlockprocessor(md.parser, self.abbrs, self.getConfigs()), 'abbr', 16)
```
Not sure why configs are being passed here as they are not used by the blockprocessor.

nbanyan replied:
They were used for `use_last_abbr` and are no longer needed.

waylan commented on markdown/extensions/abbr.py:
```diff
@@ -69,13 +93,14 @@ def iter_element(self, el: etree.Element, parent: etree.Element | None = None) -
-                abbr.tail = text[m.end():]
-                el.insert(0, abbr)
-                text = text[:m.start()]
+                if self.abbrs[m.group(0)]:
```
Curious why you felt the need to wrap this in an if statement. Seem reasonable to expect that if we have a match this should always be True. If there is a specific case where it is not, then we need a test case for it.

nbanyan replied:
This handles the situation in `test_abbr_empty()` of declaring an abbreviation without a definition. Without that test it adds `abbr` tags with an empty title instead of leaving it unmodified.

waylan replied:
Ah, right. I wasn't paying attention to what you were actually testing. You are checking the value, not the key. And I suppose a user could pass in a glossary with keys which contain empty values. Yes, this needs to stay here.

waylan commented on markdown/extensions/abbr.py:
```diff
@@ -102,7 +129,7 @@ class AbbrBlockprocessor(BlockProcessor):
     RE = re.compile(r'^[*]\[(?P<abbr>[^\\]*?)\][ ]?:[ ]*\n?[ ]*(?P<title>.*)$', re.MULTILINE)
 
-    def __init__(self, parser: BlockParser, abbrs: dict):
+    def __init__(self, parser: BlockParser, abbrs: dict, config: dict):
```
Again, this appears to be not needed.

waylan commented on tests/test_syntax/extensions/test_abbr.py:
```diff
@@ -119,21 +120,67 @@ def test_abbr_multiple_nested(self):
         )
 
-    def test_abbr_override(self):
+    def test_abbr_glossary(self):
```
I understand why the new test was added, but why was the existing test removed/replaced? Shouldn't we have both tests?

nbanyan replied:
Restored. I forgot that I didn't add that test and accidentally removed it when removing artifacts from `use_last_abbr`.

nbanyan committed:
Some linting and cleanup around abbr extension

Completed removal of artifacts from removing
`use_last_abbr` and updated documentation.

waylan commented on tests/test_syntax/extensions/test_abbr.py:
```diff
@@ -383,6 +446,47 @@ def test_abbr_with_attr_list(self):
+                <p>abbr</p>\n<p>Testing document text.</p>
+                """
+            )
+        )
```
I don't think this is the correct behavior for this test case. The syntax for definitions are modeled after reference style links. Notice how an empty link reference behaves:

```markdown
[empty link reference][empty]

[empty]: 
```

The above renders as

```html
<p>[empty link reference][empty]</p>
<p>[empty]: </p>
```

Specifically, an empty link reference is not recognized as one at all. I think this is probably how abbreviation definitions should behave as well. The fact that a test for an empty definition didn't exist suggests that this edge case wasn't given any consideration in the past. Thanks for catching another bug.

What this means is that, given the current input, the expected output for `test_abbr_empty()` would be:

```html
<p><abbr title="Abbreviation Definition">abbr</abbr></p>
<p>*[abbr]:</p>
<p>Testing document text.</p>
```

Of course, that would also mean that your attempt to create a way to override/remove a previously defined abbreviation would no longer work. Perhaps a reasonable workaround would be to add support for wrapping the definition in quotes. Then empty quotes would be parsed as an empty definition.

```markdown
*[not-a-def]:
*[empty-def]: ""
```

waylan replied:
On the other hand, there is no mechanism to predefine link references. However, with this change we now have a way to predefine abbr definitions. Therefore, it is reasonable to expect to be able to define  an empty definition to override/remove it. Therefore, maybe we can just leave things as-is. I'll need to think some more about this.

nbanyan replied:
I like the more declarative nature of the `*[empty-def]: ""` syntax.
Potential alternatives to `""`: `[]`, `{}`, `_`, `...`

There are likely others and I don't know which ones feel more Markdown-ish.

waylan replied:
Quotes are used in both inline and reference style [links](https://daringfireball.net/projects/markdown/syntax#link) for titles.  In reference style links the title can also be wrapped in regular parentheses (`()`), but in all my years I have never seen anyone actually use that. I suppose maybe it could be useful if the content needed to include quote characters. However, here we only need the wrapping characters when the content is empty, so that would be less of a concern. I think using quotes (supporting both single and double) makes the most sense in this context.

nbanyan replied:
Added handling for ignoring `*[not-a-def]:` and removing definitions with `*[empty-def]: ""` or `*[empty-def]: ''`
Also, while testing I found that `*[]: Bad Definition` resulted in abbr tags everywhere, so added testing to avoid that.

nbanyan committed:
Added proper syntax for disabling an abbreviation

Also added testing to avoid having an abbreviation
where `{"": "Definition"}`.

nbanyan committed:
Added a small bit to test_abbr.py

waylan commented:
@nbanyan thanks for the work on this. Much appreciated.
